apps/examples/camera/camera_main.c

```diff
@@ -163,6 +193,149 @@ struct camera_ctrls ctrls[] = {
 /****************************************************************************
  * Private Functions
  ****************************************************************************/
+static void circ_bbuf_deinit(struct circ_buf *c)
+{
+	sem_destroy(&c->sem);
+}
+
+static void circ_bbuf_init(struct circ_buf *c)
+{
+	int i = 0;
+	struct buf_object buf = { 0, };
+	c->head = -1;
+	c->tail = -1;
+	c->maxlen = VIDEO_BUFNUM;
+	while (i < c->maxlen) {
+		c->bobj[i] = buf;
+		i++;
+	}
+	sem_init(&c->sem, 0, 1);
+}
+
+static int circ_bbuf_push(struct circ_buf *c, struct buf_object buf)
+{
+	sem_wait(&c->sem);
+	if ((c->head == 0 && c->tail == c->maxlen - 1) || (c->tail == (c->head - 1) % (c->maxlen - 1))) {
+		sem_post(&c->sem);
+		return -1;
+	}
+	/* Insert First Element */
+	else if (c->head == -1) {
+		c->head = c->tail = 0;
+		c->bobj[c->tail] = buf;
+	} else if (c->tail == c->maxlen - 1 && c->head != 0) {
+		c->tail = 0;
+		c->bobj[c->tail] = buf;
+	} else {
+		c->tail++;
+		c->bobj[c->tail] = buf;
+	}
+	sem_post(&c->sem);
+	return 0;
+}
+
+static int circ_bbuf_pop(struct circ_buf *c, struct buf_object **buf)
+{
+	sem_wait(&c->sem);
+	if (c->head == -1) {
+		sem_post(&c->sem);
+		return -1;
+	}
+	*buf = &c->bobj[c->head];
+	if (c->head == c->tail) {
+		c->head = -1;
+		c->tail = -1;
+	} else if (c->head == c->maxlen - 1) {
+		c->head = 0;
+	} else {
+		c->head++;
+	}
+	sem_post(&c->sem);
+	return 0;
+}
+
+static pthread_addr_t write_thread(pthread_addr_t arg)
```

Comment:
I think callback mechanism is more good.
You can make thread in video upper layer and function to register callback from app or MQ, when lower layer finished capture frames, upper will get data.
Then callback or mq will send to apps.

Reply:
Current video driver use a circular buffer to hold the frames, application de-queue one at a time and process it. Since the lower camera driver get frames every 33ms, it will overwrite the buffers which are not de-queued by application. If we use callback or mq at video driver to deliver the data to application, at any given point of time there won't be any buffers available for lower camera/uvc driver to stream the data. The approach used here is same as V4L2.